# Material de aula - Redes Neurais e Deep Learning

In [1]:
#Neste versão o bias não é incluido dentro do vetor de pesos e os calculos são separados

## MLP e Aproximação de Funções - Laboratório 01

In [11]:
#CÉLULA MLP-LIB-01
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline
import math

In [12]:
#CÉLULA MLP-LIB-02
def loadDataFile(filename) :
    '''
    Esta função faz o carregamento de dados de um arquivo separado por virgulas (csv)
    
    Parametros
    -----------
    filename -> nome do arquivo de dados em formato csv
    
    
    Retorno
    -----------  
    dataframe contendo os dados do arquivo formatados numericamente.
    '''
    
    dfFile = pd.read_csv(filename, sep=';', decimal=',', thousands='.')
    return dfFile

# 1 - Implementação das funções de ativação dos neuronios do MLP

Utilizando Jupyter e NumPy: para o problema apresentado, construa duas funções para o cálculo da saída dos neurônios do MLP, uma utilizando a função sigmoide e outra utilizando a tangente hiperbolica.

### Função Sigmóide
<center>
$ \Huge f(z) = \frac{1}{1+e^{-z}} $
</center>

### Função Tangente hiperbólica
<center>
$\Huge f(z) = \frac{e^z - e^{-z}}{e^z+e^{-z}}$
</center>


In [15]:

#CÉLULA MLP-LIB-03
def sigmoid(z) :
    '''
    Esta função faz o cálculo da função de ativação do tipo sigmoide
    
    Parametros
    -----------
    z -> valor do parametro a ser calculado em f(z)
    
    Retorno
    -----------  
    valor da função sigmoide correspondente ao parametro z
    '''
    #IMPLEMENTE O CÓDIGO AQUI
    
    return 1 / (1 + np.exp(-z))

In [16]:
sigmoid(np.array([0, 0.5, 1, 0.333, 0.888]))

array([0.5       , 0.62245933, 0.73105858, 0.58248914, 0.70847727])

### Validação
<blockquote>
sigmoid(np.array([0, 0.5, 1, 0.333, 0.888]))
</blockquote>

### Resultado esperado:
<center>
array([0.5       , 0.62245933, 0.73105858, 0.58248914, 0.70847727])
</center>

In [19]:
#CÉLULA MLP-LIB-04
def tanh_function(z) :
    '''
    Esta função faz o cálculo da função de ativação do tipo tangente hiperbolica
    
    Parametros
    -----------
    z -> valor do parametro a ser calculado em f(z)
    
    Retorno
    -----------  
    valor da função tangente hiperbolica correspondente ao parametro z
    '''
    #IMPLEMENTE O CÓDIGO AQUI
    return ((np.exp(z) - np.exp(-z))/(np.exp(z) + np.exp(-z)))

In [20]:
tanh_function(np.array([-2, -1, -0.5, 0, 0.5, 1, 2]))

array([-0.96402758, -0.76159416, -0.46211716,  0.        ,  0.46211716,
        0.76159416,  0.96402758])

### Validação
<blockquote>
tanh_function(np.array([-2, -1, -0.5, 0, 0.5, 1, 2]))
</blockquote>

### Resultado esperado:
<center>
array([-0.96402758, -0.76159416, -0.46211716,  0.        ,  0.46211716,
        0.76159416,  0.96402758])
</center>

# 3 - Implementação do algoritmo de treinamento do MLP

Utilizando Jupyter e NumPy: para o problema apresentado, complete o código para construir as funções de treinamento do MLP.

## 3.1 - Implemente o código que realiza o forward step do MLP

In [21]:
#CÉLULA MLP-LIB-04
def forward_step(vInputs, vWeights, vBias) : 
    '''
    Processa o forward step
    
    Parametros
    -----------
    vInputs  -> vetor/matriz com as entradas
    vWeights -> vetor/matriz com os vetores de pesos (é um vetor de matrizes)
    vBias    -> vetor/matriz com os vetores de bias  (é um vetor de matrizes)
    
    Retorno
    -------
    Este método retorna uma tupla contendo a saída das camadas do MLP
    (Y_1, Y_2, Y_3)
    '''
    
    W_1 = vWeights[0]
    W_2 = vWeights[1]
    W_3 = vWeights[2]
    
    B_1 = vBias[0]
    B_2 = vBias[1]
    B_3 = vBias[2]
    
    #FORWARD STEP
    #Calcula saída do MLP para todas as amostras, de uma vez.
    Y_0 = vInputs #so para ficar uniforme a nomenclatura dos termos
    
    #IMPLEMENTE O CÓDIGO AQUI - INICIO
    I_1 = W_1 @ Y_0 - B_1
    Y_1 = tanh_function(I_1)
    
    I_2 = W_2 @ Y_1 - B_2
    Y_2 = tanh_function(I_2)
    
    I_3 = W_3 @ Y_2 - B_3
    Y_3 = tanh_function(I_3)
    
    #IMPLEMENTE O CÓDIGO AQUI - FIM

    #FIM DO FORWARD STEP
    
    return (Y_1, Y_2, Y_3)

In [22]:
n1 = 3 
n2 = 2 
n3 = 1 
vInputs = np.array([[0.3, 0.7]]) 
B_1 = np.array([[0.2],[0.3],[0.4]]) 
B_2 = np.array([[-0.7],[-0.3]]) 
B_3 = np.array([[0.1]]) 
W_1 = np.array([[0.4, 0.5],[0.6, 0.7],[0.8, 0.3]]) 
W_2 = np.array([[0.6, 0.2, 0.2],[0.7, 0.2, 0.8]]) 
W_3 = np.array([[0.8, 0.5]]) 
(Y_1, Y_2, Y_3) = forward_step(vInputs.T, [W_1, W_2, W_3], [B_1, B_2, B_3]) 
print('Y1={}'.format(Y_1)) 
print('--------------') 
print('Y2={}'.format(Y_2)) 
print('--------------') 
print('Y3={}'.format(Y_3)) 
print('--------------')

Y1=[[0.26362484]
 [0.35399171]
 [0.04995837]]
--------------
Y2=[[0.73474632]
 [0.53369845]]
--------------
Y3=[[0.63791268]]
--------------


### Validação
<blockquote>
    n1 = 3
    <br>
    n2 = 2
    <br>
    n3 = 1
    <br>
    vInputs = np.array([[0.3, 0.7]])
    <br>
    B_1 = np.array([[0.2],[0.3],[0.4]])
    <br>
    B_2 = np.array([[-0.7],[-0.3]])
    <br>
    B_3 = np.array([[0.1]])
    <br>
    W_1 = np.array([[0.4, 0.5],[0.6, 0.7],[0.8, 0.3]])
    <br>
    W_2 = np.array([[0.6, 0.2, 0.2],[0.7, 0.2, 0.8]])
    <br>
    W_3 = np.array([[0.8, 0.5]])
    <br>
    (Y_1, Y_2, Y_3) = forward_step(vInputs.T, [W_1, W_2, W_3], [B_1, B_2, B_3])
    <br>
    print('Y1={}'.format(Y_1))
        <br>
    print('--------------')
        <br>
    print('Y2={}'.format(Y_2))
        <br>
    print('--------------')
        <br>
    print('Y3={}'.format(Y_3))
        <br>
    print('--------------')
</blockquote>

### Resultado esperado:
<center>
    Y1=[[0.26362484]
     [0.35399171]
     [0.04995837]]
    <br>
    --------------
    <br>
    Y2=[[0.73474632]
     [0.53369845]]
    <br>
    --------------
    <br>
    Y3=[[0.63791268]]
    <br>
    --------------
</center>

## 3.2 - Implemente o código que calcula o Erro Quadrático Médio do MLP

In [27]:
#CÉLULA MLP-LIB-05
def calcMSE(vInputs, vOutputs, vWeights, vBias) : 
    '''
    Calcula o Erro Quadratico Medio (Mean Squared Error) do MLP
    
    Parametros
    -----------
    vInputs  -> vetor/matriz com as entradas
    vOutputs -> vetor/matriz com os vetores dos dados
    vWeights -> vetor/matriz com os vetores de pesos (é um vetor de matrizes)
    vBias    -> vetor/matriz com os vetores de bias  (é um vetor de matrizes)
    
    Retorno
    -------
    Este método retorna o valor de MSE para o conjunto de dados e pesos do MLP
    '''
    
    (_, _,Y_3) = forward_step(vInputs, vWeights, vBias)
    nSamples = vInputs.shape[1]

    #IMPLEMENTE O CÓDIGO AQUI - INICIO
    E_k = 1/2*(vOutputs - Y_3) ** 2
    MSE = np.sum(E_k, axis = 1)/nSamples
        
    #IMPLEMENTE O CÓDIGO AQUI - FIM
        
    return np.asscalar(MSE)

In [28]:
vInputs = np.array([[0.3, 0.7], [0.5, 0.9], [-0.3, -0.7], [-0.5, -0.9]]) 
B_1 = np.array([[0.2],[0.3],[0.4]]) 
B_2 = np.array([[-0.7],[-0.3]]) 
B_3 = np.array([[0.1]]) 
W_1 = np.array([[0.4, 0.5],[0.6, 0.7],[0.8, 0.3]]) 
W_2 = np.array([[0.6, 0.2, 0.2],[0.7, 0.2, 0.8]]) 
W_3 = np.array([[0.8, 0.5]]) 
vOutputs = np.array([[-1], [-1], [1], [1]]) 
calcMSE(vInputs.T, vOutputs.T, [W_1, W_2, W_3], [B_1, B_2, B_3])

1.2093873500459282

### Validação
<blockquote>
    vInputs = np.array([[0.3, 0.7], [0.5, 0.9], [-0.3, -0.7], [-0.5, -0.9]]) 
    <br>
    B_1 = np.array([[0.2],[0.3],[0.4]]) 
    <br>
    B_2 = np.array([[-0.7],[-0.3]]) 
    <br>
    B_3 = np.array([[0.1]]) 
    <br>
    W_1 = np.array([[0.4, 0.5],[0.6, 0.7],[0.8, 0.3]]) 
    <br>
    W_2 = np.array([[0.6, 0.2, 0.2],[0.7, 0.2, 0.8]]) 
    <br>
    W_3 = np.array([[0.8, 0.5]])
    <br>
    vOutputs = np.array([[-1], [-1], [1], [1]])
    <br>
    calcMSE(vInputs.T, vOutputs.T, [W_1, W_2, W_3], [B_1, B_2, B_3])
</blockquote>

### Resultado esperado:
<center>
    1.209387350045928
</center>